### ISSUE TITLE
interval-bitor  may be too conservative

### ISSUE THREAD

**Issue by timor** [factor/factor#2170]
Please correct me if I'm wrong, but shouldn't

```factor
IN: scratchpad auto-use 0 255 [a,b] dup interval-bitor .
T{ interval { from { 0 t } } { to { 256 t } } }
```
return `T{ interval { from { 0 t } } { to { 255 t } } }` instead?

**Comment by timor** [comment]
I think the following should work:

```factor
<PRIVATE
! Return the weight of the MSB.  For signed numbers, this does not mean the sign
! bit.
: bit-weight  ( n -- m )
    dup 0 > [ 1 + ] [ neg ] if next-power-of-2 ;

: bounds ( interval -- lower upper )
    interval>points [ first ] bi@ ;

: interval-negative? ( interval -- ? )
    to>> first 0 < ;

PRIVATE>

! In case both intervals are positive:
! lower(i1 bitor i2) = max(lower(i1),lower(i2)) (you cannot take bits away with or)
! max(i1 bitor i2) = 2 ^ max(bit-length(i1), bit-length(i2)) - 1
! In case both intervals are negative:
! max(i1 bitor i2) = -1
! lower(i1 bitor i2) = max(lower(i1),lower(i2))
! In case one is negative and the other positive, simply assume the whole bit-range.
: stricter-interval-bitor ( i1 i2 -- i3 )
    [| i1 i2 |
     i1 bounds :> ( lower1 upper1 )
     i2 bounds :> ( lower2 upper2 )
        { { [ i1 i2 [ interval-nonnegative? ] both? ]
            [ lower1 lower2 max upper1 upper2 max bit-weight 1 - [a,b] ] }
          { [ i1 i2 [ interval-negative? ] both? ]
            [ lower1 lower2 max -1 [a,b] ] }
          [
              lower1 lower2 min bit-weight upper1 upper2 max bit-weight max
              [ neg ] [ 1 - ] bi [a,b] ]
        } cond
    ] do-empty-interval ;
```

**Comment by kusumotonorio** [comment]
I'm sorry to ask you a silly question now, but I'd like to know if you can.
I am wondering about `bit-weight` in `interval-bittor`.
```Factor
! Return the weight of the MSB.  For signed numbers, this does
! not mean the sign bit.
: bit-weight  ( n -- m )
    dup [ -1/0. = ] [ 1/0. = ] bi or
    [ drop 1/0. ]
    [ dup 0 > [ 1 + ] [ neg ] if next-power-of-2 ] if ;
```
Try this:
```Factor
IN: scratchpad { 0b0001 0b0010 0b0100 0b1000 } [ bit-weight ] map .
{ 2 4 8 16 }
IN: scratchpad { 1 2 4 8 } [ bit-weight ] map .
{ 2 4 8 16 }
```
If `bit-weight` return "the weight of the MSB", why isn't the answer `{ 1 2 4 8 }` ?

**Comment by mrjbq7** [comment]
You mean something more like ``math.bitwise:bit-length``:

```factor
IN: scratchpad USE: math.bitwise
IN: scratchpad { 1 2 4 8 } [ bit-length ] map .
{ 1 2 3 4 }
```

Well, the only thing I can think of is this is basically ``bit-length 2^``, which may have some other meaning in the context of the interval algorithm... not sure.

**Comment by kusumotonorio** [comment]
The definition of `interval-bitor` is:
```Factor
: interval-bitor ( i1 i2 -- i3 )
    [
        { { [ 2dup [ interval-nonnegative? ] both? ]
            ! FIXME: this should maybe be bitweight 1 -
            [ [ max-lower-bound ] [ max-upper-bound ] 2bi bit-weight [a,b] ] }
          { [ 2dup [ interval-negative? ] both? ]
            [ max-lower-bound -1 [a,b] ] }
          [ interval-union interval-bit-weight [ neg ] [ 1 - ] bi [a,b] ]
        } cond
    ] do-empty-interval ;
```
With that in mind, I did the same thing without using this word.
```Factor
IN: scratchpad 0 255 [a,b] dup [ max-lower-bound ] [ max-upper-bound ] 2bi


--- Data stack:
0
255
IN: scratchpad 2dup bit-weight [a,b] .
T{ interval { from { 0 t } } { to { 256 t } } }

--- Data stack:
0
255
IN: scratchpad [a,b] .
T{ interval { from { 0 t } } { to { 255 t } } }
IN: scratchpad 255 bit-weight .
256
```
So if `255 bit-weight` had returned` 255 `, I should have got the expected value, `T{ interval { from { 0 t } } { to { 255 t } } }`.

```Factor
IN: scratchpad 0b10000000 bit-weight .
256
```
However, for 255 (= 2^7 + 2^6 + 2^5 + 2^4 + 2^3 + 2^2 + 2^1 + 2^0) the most significant bit has a weight of 2^7, or 128. If `255 bit-weight` returns `255`, it is semantically incorrect.

**Comment by mrjbq7** [comment]
Right, although I'm not sure that's correct either -- or at least, I need to think about it.

Here's my "revert" to the code from the original patch earlier in the issue.

If you make it return ``255``, it feels more correct, but causes problem in the compiler.

e256a4ba18fbb7916db2a22ead55cbcbf7cf920d

**Comment by kusumotonorio** [comment]
I think something needs to be fixed, but the problem is that I'm not sure if I should give priority to the `bit-weight`'s name and the intent of the action in the comment, or if I'm going to keep the  actual action and change the name.
Since `bit-weight` is used in many places, we will have to consider their impact.

**Comment by kusumotonorio** [comment]
By the way, the comment at the beginning of `intervals.factor` says:
```Factor
! Copyright (C) 2007, 2009 Slava Pestov, Doug Coleman.
! See http://factorcode.org/license.txt for BSD license.
! Based on Slate's src/unfinished/interval.slate by Brian Rice.
```
Understanding what ` src/unfinished/interval.slate` is and what it is may determine what action to take. Do you know what this is?

**Comment by kusumotonorio** [comment]
I found it.
https://github.com/briantrice/slate-language/blob/master/src/unfinished/interval.slate

**Comment by timor** [comment]
If I remember correctly, the intention of `bit-weight` was to return what the unsigned value would be if only the MSB was present.  But then wherever it is used 1 is subtracted.  So it could be that it should return e.g. `128` for `255 bit-weight` to fit the naming, but wherever it is used we would have to do a `2 * 1 -`. (I may have to check whether e256a4ba18fbb7916db2a22ead55cbcbf7cf920d changed anything regarding the original intention)

I remember that it was a bit tricky dealing with positive and negative numbers in the correct way.

The important thing was that `interval-bitor` should never be allowed to grow during compilation when repeatedly applied to integer numbers which can be represented with the same maximum number of bits (e.g. fixnums), regardless of being negative or positive.

**Comment by timor** [comment]
I thought I fixed that in #2172, but it seems to be a problem (still/again?):

```lang=factor
Factor 0.99 x86.64 (1941, heads/master-7999e72aec, Jan  1 1970 00:00:01)
[GCC 8.3.0] on linux

IN: scratchpad auto-use USING: compiler.test kernel.private ;
IN: scratchpad auto-use [ { fixnum fixnum } declare [ 0 255 clamp ] bi@ bitor ] final-info first interval>> .
T{ interval { from { 0 t } } { to { 256 t } } }
```

**Comment by kusumotonorio** [comment]
This may be because the current `interval-bitor` has been changed so that 1 is not subtracted from the result of `bit-weight`.

That's because there are compelling reasons. We need to think about why there are problematic side effects.

**Comment by timor** [comment]
> That's because there are compelling reasons.

I would be interested in those.  Do you think that the original behavior:

> ```factor
> IN: scratchpad auto-use 0 255 [a,b] dup interval-bitor .
> T{ interval { from { 0 t } } { to { 256 t } } }
> ```

is actually correct?  Or are you talking about `bit-weight` itself?  `bit-weight` is really just a helper word that I wrote to fix above behavior, and did not serve any other purpose at the time.

> Since `bit-weight` is used in many places, we will have to consider their impact.

As far as I can tell, it is only used in `intervals.factor` and `intervals-tests.factor`, and only for the sake of implementing the three bitwise interval operations.

**Comment by kusumotonorio** [comment]
> That's because there are compelling reasons.

This means that @mrjbq7 changed it at e256a4b.

>  Do you think that the original behavior:
```Factor
IN: scratchpad auto-use 0 255 [a,b] dup interval-bitor .
T{ interval { from { 0 t } } { to { 256 t } } }
```

Certainly, this should be `T{ interval { from { 0 t } } { to { 255 t } } }`.
It seems to me that the changes to make that happen are:
```Factor
: interval-bitxor ( i1 i2 -- i3 )
     ! Inaccurate.
     [
        2dup [ interval-nonnegative? ] both?
         [
             [ interval>points [ first ] bi@ ] bi@
             4array supremum 0 swap >integer next-power-of-2 1 - [a,b] ! <----
         ] [ 2drop [-inf,inf] ] if
    ] do-empty-interval ;
```

The operation for the interval including the negative region is simplified in the original. That makes sense to me. This is because, unlike add, subtract and so on, signs are generally not taken into account in Boolean operations.

**Comment by kusumotonorio** [comment]
```Factor
IN: scratchpad 256 next-power-of-2 .
256
```
Hmm, my code is bad.　In this case one should not be subtracted.

**Comment by kusumotonorio** [comment]
Oh, I see. I understand why 1 is added at `bit-weight`.

`bit-weight 1 -` and `1 dup dup next-power-of-2 dup rot = [ swap 1 - + ] [ nip 1 - ] if .`  which I improvised seem to give the same answer.

**Comment by kusumotonorio** [comment]
So `bit-weight  ( n -- m )` outputs the smallest power of 2 greater than n, right?

**Comment by timor** [comment]
If you assume a number to be unsigned, then yes.  If you assume a number to be signed, then it is the power of two which the sign bit "occupies".

So under the current behavior: `42 bit-weight` and `-42 bit-weight` return the same result: `64`.  This is because any bit-wise operations on these numbers will, when the result is interpreted as an unsigned number, never be larger than this.  I think.  It's been a while...

**Comment by kusumotonorio** [comment]
Thanks, @timor . 

Before adopting this change, `interval-bitor` and `interval-bitxor` had  comments : "Inaccurate.". I understand that these comments are no longer needed because they are more accurate.
`Interval-shift` also has the comment "Inaccurate; could be tiger". Does this have anything to do with the optimization problem (e256a4b) that arise from adapting changes?

**Comment by timor** [comment]
> Does this have anything to do with the optimization problem ([e256a4b](https://github.com/factor/factor/commit/e256a4ba18fbb7916db2a22ead55cbcbf7cf920d)) that arise from adapting changes?

I don't think so.  Blame information traces that back to 7ffd9c95baf0fa58de7a01711074c0823b424d04.

However, I am not familiar with what exactly the problem is in e256a4b.

**Comment by mrjbq7** [comment]
This has been fixed since 971b5660f9a822f439d158fbff11b66177a6a91d.
